rllib/algorithms/appo/tf/appo_tf_learner.py

```diff
@@ -61,17 +61,39 @@ def compute_loss_for_module(
             trajectory_len=hps.rollout_frag_or_episode_len,
             recurrent_seq_len=hps.recurrent_seq_len,
         )
-        values_time_major = make_time_major(
-            values,
+        rewards_time_major = make_time_major(
+            batch[SampleBatch.REWARDS],
             trajectory_len=hps.rollout_frag_or_episode_len,
             recurrent_seq_len=hps.recurrent_seq_len,
         )
-        bootstrap_value = values_time_major[-1]
-        rewards_time_major = make_time_major(
-            batch[SampleBatch.REWARDS],
+        values_time_major = make_time_major(
+            values,
             trajectory_len=hps.rollout_frag_or_episode_len,
             recurrent_seq_len=hps.recurrent_seq_len,
         )
+        bootstrap_values_time_major = make_time_major(
+            batch[SampleBatch.VALUES_BOOTSTRAPPED]
+        )
+        # Then add the shifted-by-one bootstrapped values to that to yield the final
+        # value tensor. Use the last ts in that resulting tensor as the
+        # "bootstrapped" values for vtrace.
+        shape = tf.shape(values_time_major)
+        B = shape[1]
+        # Augment `values_time_major` by one timestep at the end (all zeros).
+        values_time_major = tf.concat([values_time_major, tf.zeros((1, B))], axis=0)
+        # Augment `bootstrap_values_time_major` by one timestep at the beginning
+        # (all zeros).
+        bootstrap_values_time_major = tf.concat(
+            [tf.zeros((1, B)), bootstrap_values_time_major], axis=0
+        )
+        # Note that the `SampleBatch.VALUES_BOOTSTRAPPED` values are always recorded
+        # ONLY at the last ts of a trajectory (for the following timestep,
+        # which is one past(!) the last ts). All other values in that tensor are
+        # zero.
+        # Adding values and bootstrap_values yields the correct values+bootstrap
+        # configuration, from which we can then take t=-1 (last timestep) to get
+        # the bootstrap_value arg for the vtrace function below.
+        values_time_major += bootstrap_values_time_major
```

Comment:
I do think we still need this. For the following reason:
In APPO and IMPALA, we build/preprocess the train batch along the rollout_fragment_len (or `max_seq_len` if LSTM) boundaries. This means that in case a rollout ends within an episode, this rollout's last trajectory will end up in the train batch with a zero-padded right side, thus, the bootstrapped value for this fragment is in the middle of the train batch, NOT at time-axis index -1!
So adding these two together here covers that particular case as well. It'll lead to most bootstrapped value to be at the end of the train batch rows, but some (in those cases where the rollout was done within an episode) will be located in the middle of the train batch's rows. Here: "row" means a trajectory (along the T-axis) within the (B, T, ...) train batch.

Reply:
I'm struggling to understand how `compute_bootstrap_value` handles this situation where the rollout ends within an episode.  I'm reading the code for it, and on the surface I don't see anything that searches in the middle of an episode for the reward at the terminated timestep. It looks like we're only checking the last timestep in the sample batch.

Reply:
Note that `compute_bootstrap_value` is only ever called by a Policy's postprocess_trajectory method, which - by design - is only called when a rollout has ended (either within or at the terminal of an episode).
If at a terminal: Assume the value to be 0.0 (no value computation necessary)
If NOT at a terminal: Use the Policy's vf to compute the value at the last timestep of the trajectory. This is the "bootstrap" value to be used in the losses (instead of dropping the last ts and using that ts as a "bootstrapped" value).